pkg/apis/internal.acorn.io/v1/memory.go

```diff
@@ -0,0 +1,122 @@
+package v1
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"cuelang.org/go/cue/literal"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+func ParseMemory(s []string) (Memory, error) {
+	result := Memory{}
+	for _, s := range s {
+		workload, memBytes, specific := strings.Cut(s, "=")
+
+		// If setting all, swap workload and memBytes
+		if !specific {
+			memBytes = workload
+			workload = ""
+		}
+
+		quantity, err := ParseQuantityString(memBytes)
+		if err != nil {
+			return Memory{}, err
+		}
+
+		result[workload] = &quantity
+	}
+	return result, nil
+}
+
+func ParseQuantityString(memory string) (int64, error) {
+	numInfo := literal.NumInfo{}
+	err := literal.ParseNum(memory, &numInfo)
+	if err != nil {
+		return -1, err
+	}
+
+	quantity, err := strconv.ParseInt(numInfo.String(), 10, 64)
+	if err != nil {
+		return -1, err
+	}
+
+	return quantity, nil
+}
+
+var (
+	ErrInvalidAcornMemory   = errors.New("invalid memory from acorn")
+	ErrInvalidSetMemory     = errors.New("invalid memory set by user")
+	ErrInvalidDefaultMemory = errors.New("invalid memory default")
+)
+
+type MemoryError struct {
+	Err      error
+	Quantity string
+}
+
+func MemoryToRequirements(specMem Memory, containerName string, container Container, specMemDefault, specMemMaximum *int64) (*corev1.ResourceRequirements, MemoryError) {
+	var memMaximum, memDefault int64
+	if specMemDefault != nil {
+		memDefault = *specMemDefault
+	}
+	if specMemMaximum != nil {
+		memMaximum = *specMemMaximum
+	}
+
+	// Determine which memory should be used to set the resource limit/requests. Gets set
+	// 4 ways: User setting a specific workload, user setting all workloads, Acornfile, or
+	// from the apiv1.Config default.
+	memBytes, err := memDefault, ErrInvalidDefaultMemory
+	if m, set := specMem[containerName]; set && m != nil {
+		err = ErrInvalidSetMemory
+		memBytes = *m
+	} else if specMem[""] != nil {
+		err = ErrInvalidSetMemory
+		memBytes = *specMem[""]
+	} else if container.Memory != nil {
+		err = ErrInvalidAcornMemory
+		memBytes = *container.Memory
+	}
+
+	// For maximum memory, 0 is equivalent to "unlimited"
+	if memMaximum != 0 {
+		var (
+			maxQuantity     = resource.NewQuantity(memMaximum, resource.BinarySI).String()
+			defaultQuantity = resource.NewQuantity(memBytes, resource.BinarySI).String()
+			bytesQuantity   = resource.NewQuantity(memBytes, resource.BinarySI).String()
```

Comment:
small: looks like defaultQuantity and bytesQuantity are being set to same value here maybe eliminate one call to NewQuantity()